src/base/torrentfileswatcher.h

```diff
@@ -0,0 +1,132 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#pragma once
+
+#include <QHash>
+#include <QSet>
+
+#include "base/bittorrent/addtorrentparams.h"
+
+class QFileSystemWatcher;
+class QThread;
+class QTimer;
+
+namespace BitTorrent
+{
+    class MagnetUri;
+}
+
+class TorrentFilesWatcherWorker;
+
+/*
+ * Watches the configured directories for new .torrent files in order
+ * to add torrents to BitTorrent session. Supports Network File System
+ * watching (NFS, CIFS) on Linux and Mac OS.
+ */
+class TorrentFilesWatcher final : public QObject
+{
+    Q_OBJECT
+    Q_DISABLE_COPY(TorrentFilesWatcher)
+
+public:
+    struct WatchedFolderOptions
+    {
+        BitTorrent::AddTorrentParams addTorrentParams;
+        bool recursive = false;
+    };
+
+    static void initInstance();
+    static void freeInstance();
+    static TorrentFilesWatcher *instance();
+
+    static QString makeCleanPath(const QString &path);
+
+    QHash<QString, WatchedFolderOptions> folders() const;
+    void setWatchedFolder(const QString &path, const WatchedFolderOptions &options);
+    void removeWatchedFolder(const QString &path);
+
+signals:
+    void watchedFolderSet(const QString &path, const WatchedFolderOptions &options);
+    void watchedFolderRemoved(const QString &path);
+
+private slots:
+    void onMagnetFound(const BitTorrent::MagnetUri &magnetURI, const BitTorrent::AddTorrentParams &addTorrentParams);
+    void onTorrentFound(const BitTorrent::TorrentInfo &torrentInfo, const BitTorrent::AddTorrentParams &addTorrentParams);
+
+private:
+    explicit TorrentFilesWatcher(QObject *parent = nullptr);
+    ~TorrentFilesWatcher() override;
+
+    void load();
+    void loadLegacy();
+    void store();
+
+    void doSetWatchedFolder(const QString &path, const WatchedFolderOptions &options);
+
+    static TorrentFilesWatcher *m_instance;
+
+    QHash<QString, WatchedFolderOptions> m_watchedFolders;
+
+    QThread *m_ioThread = nullptr;
+    TorrentFilesWatcherWorker *m_asyncWorker = nullptr;
+};
+
+class TorrentFilesWatcherWorker final : public QObject
```

Comment:
Does it make sense to move it to separate files? Or, for example, put only its declaration in a separate header?

Reply:
@glassez 

I vote separate files, I think it stands on its own.

Reply:
>I think it stands on its own.

No. It's intended only for private use by `TorrentFilesWatcher` class. Otherwise I would have no such question.

Reply:
why not move it to .cpp file?

Reply:
> No. It's intended only for private use by `TorrentFilesWatcher` class. Otherwise I would have no such question.

That doesn't mean it can't be in its own cpp file. Otherwise I would have no such answer... But to me it seems that you've already decided what you think is best :).

Reply:
>why not move it to .cpp file?

At least it requires `#include "torrentfileswatcher.moc"` there (due to signals in this class). Do we accept it? If yes, what exact the place I can put it (IIRC, it should be put after class declarations).

Reply:
>That doesn't mean it can't be in its own cpp file. 

I agree. You can even put the definitions of different methods of the same class in separate files. The question is whether it makes sense. In some specific cases, it does.
Here, I perceive it inseparably from the `TorrentFilesWatcher` class (only as a detail of its implementation).
What's really not good is having this declaration here. So I have to either move it to a separate header that will be included directly in the \.cpp file, or move the declaration itself in the \.cpp file.

Reply:
Made it declared private for review.